### ISSUE TITLE
Saucery3 hanging against NUnit 3.10

### ISSUE THREAD

**Issue by agray** [nunit/nunit#2481]
Saucery is now open source!

The NUnit 3 implementation is [here](https://github.com/Sauceforge/Saucery3).

We seek to release a new version of Saucery for each release of NUnit. Currently we are depending on NUnit 3.8.1 and seek to use the newest available NUnit Console Runner (currently 3.7).

The unit tests execute but then our Selenium tests hang.  NUnit Log files attached.

    C:\Personal\Jenkins\workspace\FullCircle\Saucery\Saucery3NUnit>"C:\Program Files (x86)\NUnit.org\nunit-console\nunit3-console.exe" 
    Saucery\NUnit\Version3\Saucery3Tester\bin\Release\Saucery3Tester.dll --result:junit-selenium-
    testsuite.xml;transform=nunit3-junit.xslt --trace=Verbose  
    NUnit Console Runner 3.7.0  
    Copyright (c) 2017 Charlie Poole, Rob Prouse 
 
    Runtime Environment 
      OS Version: Microsoft Windows NT 10.0.15063.0 
      CLR Version: 4.0.30319.42000 
 
    Test Files 
      Saucery\NUnit\Version3\Saucery3Tester\bin\Release\Saucery3Tester.dll 

junit-selenium-testsuite.xml file is empty, that is why it is not attached.
 

[InternalTrace.96.log](https://github.com/nunit/nunit/files/1367659/InternalTrace.96.log)
[InternalTrace.11480.Saucery3Tester.dll.log](https://github.com/nunit/nunit/files/1367660/InternalTrace.11480.Saucery3Tester.dll.log)
[InternalTrace.11936.UnitTests.dll.log](https://github.com/nunit/nunit/files/1367661/InternalTrace.11936.UnitTests.dll.log)
[InternalTrace.11996.log](https://github.com/nunit/nunit/files/1367662/InternalTrace.11996.log)
[nunit-agent_11480.log](https://github.com/nunit/nunit/files/1367663/nunit-agent_11480.log)
[nunit-agent_11936.log](https://github.com/nunit/nunit/files/1367664/nunit-agent_11936.log)

**Comment by rprouse** [comment]
I expect this is a duplicate of our various parallelization hangs that are currently being looked at. I am going to mark this as confirm to confirm if the other fixes fix this issue.

**Comment by agray** [comment]
Thanks @rprouse, Do the attached log files show this is a parallelism bug or has issue #1118 reared its head again?  i.e. Are the tests not being discovered?

**Comment by agray** [comment]
Will try testing with 3.8.2 when it is released and report results here.

**Comment by agray** [comment]
Will try testing against the 3.9 release and report results here.

**Comment by agray** [comment]
Now that 3.9 is released I will action this during the week.

**Comment by agray** [comment]
I have upgraded Saucery3 to NUnit 3,9. Unfortunately it is still hanging.  I am not seeing any tests reach SauceLabs dashboard.  Would appreciate your assistance to work out what is the underlying cause.  Saucery has been working up to Release 3.6

I have attached the NUnit logs

[InternalTrace.5940.log](https://github.com/nunit/nunit/files/1466945/InternalTrace.5940.log)
[InternalTrace.8456.log](https://github.com/nunit/nunit/files/1466946/InternalTrace.8456.log)
[InternalTrace.10140.Saucery3Tester.dll.log](https://github.com/nunit/nunit/files/1466947/InternalTrace.10140.Saucery3Tester.dll.log)
[InternalTrace.14948.UnitTests.dll.log](https://github.com/nunit/nunit/files/1466948/InternalTrace.14948.UnitTests.dll.log)

[nunit-agent_10140.log](https://github.com/nunit/nunit/files/1466949/nunit-agent_10140.log)
[nunit-agent_14948.log](https://github.com/nunit/nunit/files/1466950/nunit-agent_14948.log)

**Comment by CharliePoole** [comment]
You show two runs here. The one against  `UnitTests.dll`, comprising processes 8456 and 14948, seems to run correctly to completion. The one against `Saucery3Tester.dll`, comprising process 5940 and 10140, is the one that hangs.

The key log is InternalTrace.10140.Saucery3Tester.dll.log. Examining it from the top, we see that after initialization, the dll itself is scheduled to run on a non-parallel worker. The included levels of tests represented by namespaces are all queued and executed on the same non-parallel worker up to the point where the four fixture instances are enqueued - two instances each of OpenSauceFixture and OpenSauceFixture2. All this happens on thread 9.

The non-parallel shift terminates and a parallel shift begins. The four fixtures begin execution on threads 11, 12, 13 and 15. If you have any one-time setup, it appears to complete, because each of them then schedules a test to run using the Direct strategy, i.e. to execute on the same thread as that used for startup. The tests are two instances each of `DoSomethingElseOnAWebPageAgainWithSelenium` and `DoSomethingElseOnAWebPageAgainWithSelenium2`. At that point, your tests hang.

I would conclude that something in those four tests creates a deadlock so that none of them can complete.

**Comment by agray** [comment]
The tests are clicking on a link (at most) on a guinea pig web page (supplied by SauceLabs) and asserting some content on the page.  I don't see any source of a deadlock.  The code is now open source. [Here](https://github.com/Sauceforge/Saucery3/tree/master/Saucery3Tester/Tests).

**Comment by CharliePoole** [comment]
I can't see anything obvious but I notice you have a base class where something may be going on. All the log tells us is that NUnit launched your four tests on four threads. We don't know if they hang in the setup, test method or TearDown only that all activity stops at that point as far as NUnit is concerned. Try running with parallelism disabled to see what happens.

**Comment by agray** [comment]
Thanks Charlie.  

If I comment out all the instances of [Parallelizable(ParallelScope.Fixtures)] as you can see in this [commit](https://github.com/Sauceforge/Saucery3/commit/82f88cf8b7ad4789cc7647afcfbed4df35d7a90a), it "works" again but it is only one test at a time.

- Enumeration of platforms by tests - WORKS
- NUnit Data Driven TestsCases - WORKS
- Parallelism - NO

What is the correct way to apply Parallelism?  

SauceryBase, SauceryAndroidBase and SauceryIOSBase would be the key classes.

**Comment by agray** [comment]
Saucery essentially uses the functionality described in [this](https://github.com/nunit/docs/wiki/TestFixtureData) nunit wiki page.  Is this "yield enumeration" compatible with NUnit 3.9's new parallel execution capabilities?

**Comment by agray** [comment]
I have tried:
1. Adding LevelOfParallelism attribute to Assembly and ParallelScope.Children on the base classes
2. Adding LevelOfParallelism attribute to Assembly and ParallelScope.Children on the classes containing tests directly.

In both cases the tests seem to hang and dont get submitted to SauceLabs.

Is there some particular way I'm supposed to add Parallelism that I haven't tried yet?

**Comment by CharliePoole** [comment]
@agray... In a conference today but I'll get back to this as soon as I can.

**Comment by agray** [comment]
Thanks @CharliePoole.

**Comment by agray** [comment]
Any thoughts on this @CharliePoole?

**Comment by agray** [comment]
@CharliePoole: Have you had a chance to look at this?

**Comment by CharliePoole** [comment]
@agray I'm back in Seattle and I'll take a closer look at this today. Here are a few notes on your accumulated comments...

* We have no experience with applying parallelism at the base class level. No attempt has ever been made to make that work, test it or even give an error in case you tried to do it. Tests and fixtures "inherit" parallelism or non-parallelism from their containers. That has been our focus.

* Inheriting from a base class is something we could consider as a feature, but I would worry about it particularly as used in a framework. The base class would have to be written so that any derived class can run in parallel and the users would need to receive instruction on what they may or may not do in the derived class.

* Remember, applying the `ParallelizableAttribute` is a promise to NUnit that the test may be run in parallel. It does nothing to __make__ the test capable of running in that way. It depends on the specific code written by the user. If we allow inheritance, then we are inheriting that promise.

* You mention you are using the functionality of `TestFixtureData` and ask how it relates to parallelism. The answer is "Not at all". Test fixtures and tests are discovered and loaded sequentially before the run ever begins. It's only at runtime that the parallel feature applies.

I'll download your code and work on it later today. Sorry you have waited so long. FYI... I suspect your mentioning me directly is keeping others from replying.

**Comment by agray** [comment]
Has anyone been able to look at this please?  Parallelism is one feature I have not been able to get working in NUnit with Saucery

**Comment by rprouse** [comment]
@agray is this still happening with 3.11 or with recent nightly builds?

**Comment by rprouse** [comment]
I am closing old confirm issues that have not had comments or made progress in several years. If anyone comes back with a compelling argument for these issues, we can reopen.
